src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,291 @@
+                {
+                    _metadataReferences = metadataReferences;
+                    _pathToIndexMap = CreatePathToIndexMap(metadataReferences);
+                    _indecesToRemove = new HashSet<int>();
+                }
+
+                private static ImmutableDictionary<string, HashSet<int>> CreatePathToIndexMap(ImmutableArray<MetadataReference> metadataReferences)
+                {
+                    var pathToIndexMap = ImmutableDictionary.CreateBuilder<string, HashSet<int>>(PathUtilities.Comparer);
+
+                    for (var index = 0; index < metadataReferences.Length; index++)
+                    {
+                        var metadataReference = metadataReferences[index];
+
+                        var filePath = GetFilePath(metadataReference);
+                        if (!pathToIndexMap.TryGetValue(filePath, out var indeces))
+                        {
+                            indeces = new HashSet<int>();
+                            pathToIndexMap.Add(filePath, indeces);
+                        }
+
+                        indeces.Add(index);
+                    }
+
+                    return pathToIndexMap.ToImmutable();
+                }
+
+                public bool Contains(string filePath) => _pathToIndexMap.ContainsKey(filePath);
+
+                public void Remove(string filePath)
+                {
+                    if (filePath != null && _pathToIndexMap.TryGetValue(filePath, out var indexSet))
+                    {
+                        _indecesToRemove.AddRange(indexSet);
+                    }
+                }
+
+                public ImmutableArray<MetadataReference> GetRemaining()
+                {
+                    var results = ImmutableArray.CreateBuilder<MetadataReference>(initialCapacity: _metadataReferences.Length);
+
+                    for (var index = 0; index < _metadataReferences.Length; index++)
+                    {
+                        if (!_indecesToRemove.Contains(index))
+                        {
+                            results.Add(_metadataReferences[index]);
+                        }
+                    }
+
+                    return results.ToImmutable();
+                }
+
+                public ProjectInfo Find(IEnumerable<ProjectInfo> projectInfos)
+                {
+                    foreach (var projectInfo in projectInfos)
+                    {
+                        if ((projectInfo.OutputRefFilePath != null && _pathToIndexMap.ContainsKey(projectInfo.OutputRefFilePath)) ||
+                            (projectInfo.OutputFilePath != null && _pathToIndexMap.ContainsKey(projectInfo.OutputFilePath)))
+                        {
+                            return projectInfo;
+                        }
+                    }
+
+                    return null;
+                }
+
+                public ProjectFileInfo Find(IEnumerable<ProjectFileInfo> projectFileInfos)
```

Comment:
> Is the assumption that projectFileInfos are just different targets of a multi-targeting project?

Yes. That is part of the fallout this change. Previously, there was a 1:1 mapping between IProjectFile and ProjectFileInfo. Now there's a 1:n mapping, where n is the number of target frameworks.

> I guess the "pick first" logic seems odd here. Or maybe "FindFirstProjectReferenced" is a bit clearer of a name.

Why is that? I'm not sure how else this could be handled. Could you elaborate?